### DISCUSSION TITLE
how can I keep a PyAny pointer in rust pyclass

### DISCUSSION THREAD

**Discussion by shenjackyuanjie** [PyO3/pyo3#2931] [Questions] [answered]
```rust

pub mod camera {
    use pyo3::prelude::*;

    #[pyclass(name = "Camera_rs")]
    pub struct CameraRs{
        pub dx: f64,
        pub dy: f64,
        #[pyo3(get, set)]
        pub zoom: f64,
        #[pyo3(get, set)]
        pub max_zoom: f64,
        #[pyo3(get, set)]
        pub min_zoom: f64,
        pub window: &PyAny,
    }

    #[pymethods]
    impl CameraRs {
        #[new]
        fn py_new(dx: f64, dy: f64, window: &PyAny) -> PyResult<Self> {
            return Ok(CameraRs {dx,
                                dy,
                                zoom: 1 as f64,
                                min_zoom: 1 as f64,
                                max_zoom: 1 as f64,
                                window})
        }

        fn __enter__(&self) -> PyResult<()> {
            println!("I enter!");
            return Ok(())
        }

        #[allow(unused_variables)]
        fn __exit__(&self, exception_type: &PyAny, exception_value: &PyAny, traceback: &PyAny) -> PyResult<()> {
            println!("I exit!");
            return Ok(())
        }
        
    }

}
```
I want to keep a window pointer in my pyclass, but rustc tells me I can't

**Comment by birkenfeld** [comment, marked as answer]
It needs to be `Py<PyAny>`. [This section of the guide](https://pyo3.rs/v0.18.0/types#object-types) may be helpful to understand the difference.

In short: `&'py PyAny` is only valid to use as long as you hold the Python Interpreter Lock (GIL). Therefore it is not safe to put into a `pyclass`, which can live indefinitely as a Python object. `Py<PyAny>` requires a `Python` GIL token to access and is therefore safe to share around.

**Reply by shenjackyuanjie** [reply]
oh! thanks, my bad
(btw, is this pyclass useable with `with`?

**Reply by birkenfeld** [reply]
To be useable with Python `with`, you just need to implement `__enter__` and `__exit__` methods, no special support is present (or necessary).

[Here](https://github.com/samuelcolvin/watchfiles/blob/afb8f34d9d6ec9eb2a33b2db53d915d56ddf6413/src/lib.rs#L308) is an example I found.

**Reply by shenjackyuanjie** [reply]
thanks!

**Reply by shenjackyuanjie** [reply]
> To be useable with Python `with`, you just need to implement `__enter__` and `__exit__` methods, no special support is present (or necessary).
> 
> [Here](https://github.com/samuelcolvin/watchfiles/blob/afb8f34d9d6ec9eb2a33b2db53d915d56ddf6413/src/lib.rs#L308) is an example I found.

wait, how can I call a self function in __enter__ ?

**Reply by shenjackyuanjie** [reply]
```rust
        /// https://github.com/PyO3/pyo3/discussions/2931#discussioncomment-4820729 for finding this
        /// https://github.com/PyO3/pyo3/issues/1205#issuecomment-1164096251 for advice on `__enter__`
        pub fn __enter__(py_self: Py<Self>) -> Py<Self> {
            py_self.call_method0("start")?;
            return py_self
        }
```
I want to call `start` over here

**Reply by birkenfeld** [reply]
To clarify, you want to call `self.start()` in Rust instead?

My example probably wasn't the best - it uses `py_self: Py<Self>` in order to save the marshaling to `self` and back when only returning `self` again. When you actually need Rust's `self`, you can just take it like in other pymethods:

```
fn __enter__(&self) -> PyResult<Self> {  // or &mut self
    self.start(...)?;
    Ok(self)
}
```

**Reply by shenjackyuanjie** [reply]
![image](https://user-images.githubusercontent.com/54507071/215662258-9268d8aa-a215-452d-ad95-13ae56e04d84.png)
![image](https://user-images.githubusercontent.com/54507071/215662536-745fa9fd-d37c-4ce6-843f-ef246d453cc7.png)

hmmmmm
rustc and rust plugin tells me I need to return `self` not `&self`

**Reply by shenjackyuanjie** [reply]
![image](https://user-images.githubusercontent.com/54507071/215663163-99165638-af14-440e-ad46-7be0e46ad5ae.png)

**Reply by birkenfeld** [reply]
Yeah sorry, that was not thought through.

Better to do stay with `PyRef`, since it de-references automatically to `Self`:
```
    fn __enter__(slf: PyRef<Self>) -> PyResult<PyRef<Self>> {
        slf.start()?;
        Ok(slf)
    }
```

**Reply by shenjackyuanjie** [reply]
ok! that work! super thanks!
